aten/src/ATen/TensorOptions.h

```diff
@@ -20,12 +20,16 @@ namespace at {
 /// `type()` to return a variable type instead of a tensor type, such that
 /// variables are created inside factory methods, instead of tensors.
 struct TensorOptions {
-  /// Constructs the `TensorOptions` with valid defaults, which are:
-  /// - dtype: float
-  /// - device: CPU
-  /// - layout: strided
+  TensorOptions() : TensorOptions(/*use_thread_local_default_options=*/true) {}
+
+  /// Constructs the `TensorOptions` with defaults taken from the thread local
+  /// `TensorOptions` object if `use_thread_local_default_options`, else
+  /// defaults to:
+  /// - dtype: kFloat,
+  /// - device: kCPU,
+  /// - layout: kStrided,
   /// - requires_grad: false
-  TensorOptions() = default;
+  explicit TensorOptions(bool use_thread_local_default_options);
```

Comment:
NB: this is a little inconsistent with how the PyTorch API works. Consider the two corresponding variants:

```
kwargs = {} # default options
# ...
t = torch.zeros(2, **kwargs)
```

versus

```
TensorOptions options(true);
// ...
auto t = at::zeros({2}, options);
```

The former reads the TLS at the time of zeros construction; the latter reads TLS at the time of options construction.

Reply:
I think my preference is, in fact, for the former semantics. But you might want to poll some other people.

Reply:
I don't think it's very common that someone will default construct a `TensorOptions` and then pass it to a function, I think 99.9% of uses cases here are that someone calls `at::zeros({...})` and then the `TensorOptions` argument defaults to the current thread local options at the time of the `zeros` call, like in Python